conans/model/build_info.py

```diff
@@ -41,49 +47,163 @@ def __init__(self):
+        element = unsorted_names[0]
+        while unsorted_names:
+            try:
+                deps = self._components[element].deps
+            except KeyError as e:
+                raise ConanException("Component %s not found in cpp_info object" % e)
+            if all(dep in [c.name for c in sorted_comps] for dep in deps):
+                sorted_comps.append(self._components[element])
+                unsorted_names.remove(element)
+                element = unsorted_names[0] if unsorted_names else None
+            else:
+                for dep in deps:
+                    if dep not in [c.name for c in sorted_comps]:
+                        element = dep
+                        break
+        return sorted_comps
+
+    @property
+    def libs(self):
+        if self._components:
+            result = []
+            for component in self._sorted_components:
+                for sys_dep in component.system_deps:
+                    if sys_dep and sys_dep not in result:
+                        result.append(sys_dep)
+                if component.lib:
+                    result.append(component.lib)
+            return result
+        else:
+            return self._libs
+
+    @libs.setter
+    def libs(self, libs):
+        assert isinstance(libs, list), "'libs' attribute should be a list of strings"
+        if self._components:
+            raise ConanException("Setting first level libs is not supported when Components are "
+                                 "already in use")
+        self._libs = libs
+
+    @property
+    def exes(self):
+        if self._components:
+            return [component.exe for component in self._sorted_components if component.exe]
+        else:
+            return self._exes
+
+    @exes.setter
+    def exes(self, exes):
+        assert isinstance(exes, list), "'exes' attribute should be a list of strings"
+        if self._components:
+            raise ConanException("Setting first level exes is not supported when Components are "
+                                 "already in use")
+        self._exes = exes
+
+    @property
+    def system_deps(self):
+        if self._components:
+            result = []
+            for component in self._sorted_components:
+                if component.system_deps:
+                    for system_dep in component.system_deps:
+                        if system_dep and system_dep not in result:
+                            result.append(system_dep)
+            return result
+        else:
+            return self._system_deps
+
+    @system_deps.setter
+    def system_deps(self, system_deps):
+        assert isinstance(system_deps, list), "'system_deps' attribute should be a list of strings"
+        if self._components:
+            raise ConanException("Setting first level system_deps is not supported when Components "
+                                 "are already in use")
+        self._system_deps = system_deps
+
+    def __getitem__(self, key):
```

Comment:
this going to be confusing with `deps_cpp_info`, as it will become something like `deps_cpp_info["boost"]["asio"]`. may be change it to the more explicit `deps_cpp_info["boost"].components["asio"]`?

Reply:
If at the `cppinfo` we are using `cpp_info["asio"]` I'm ok with `deps_cpp_info["boost"]["asio"]`